materialize-s3-parquet/driver.go

```diff
@@ -0,0 +1,343 @@
+	"lz4":    parquet.CompressionCodec_LZ4,
+	"zstd":   parquet.CompressionCodec_ZSTD,
+}
+
+func (r resource) Validate() error {
+	if r.PathPrefix == "" {
+		return fmt.Errorf("pathPrefix in a resource should not be None")
+	}
+
+	if r.CompressionType != "" {
+		if _, ok := compressionTypeToCodec[r.CompressionType]; !ok {
+			return fmt.Errorf("invalid compressionType, expecting one of %v", reflect.ValueOf(compressionTypeToCodec).MapKeys())
+		}
+	}
+
+	return nil
+}
+
+func (r resource) CompressionCodec() parquet.CompressionCodec {
+	if r.CompressionType == "" {
+		return parquet.CompressionCodec_SNAPPY
+	}
+
+	return compressionTypeToCodec[r.CompressionType]
+}
+
+// The structure of a driver checkpoint.
+type driverCheckpoint struct {
+	// The flow checkpoint (base64-encoded), which marks the txn that has been successfully
+	// materialized and stored in the cloud. If the materialization process is stopped
+	// for any reason, this is the checkpoint to resume from.
+	B64EncodedFlowCheckpoint string `json:"b64EncodedFlowCheckpoint"`
+	// The sequence number used to name the next files to be uploaded to the cloud.
+	// To be specific, the next parquet file from the i-th binding is named using the deterministic pattern of
+	// "<KeyBegin>_<KeyEnd>_<NextSeqNumList[i]>.pq".
+	// The NextSeqNumList[i] is increased by 1 after each successful upload from the i-th binding.
+	NextSeqNumList []int `json:"nextSeqNumList"`
+}
+
+func (dcp driverCheckpoint) Validate() error { return nil }
+
+// Creates a driver checkpoint, and encodes it into a json.RawMessage to populate the DriverCheckPointJson field in `prepared` response.
+func marshalDriverCheckpointJSON(flowCheckpoint []byte, nextSeqNumList []int) (json.RawMessage, error) {
+	if flowCheckpoint != nil && len(flowCheckpoint) == 0 {
+		panic("empty checkpoint received.")
+	}
+
+	dcp := &driverCheckpoint{
+		B64EncodedFlowCheckpoint: base64.StdEncoding.EncodeToString(flowCheckpoint),
+		NextSeqNumList:           nextSeqNumList,
+	}
+
+	md, err := json.Marshal(dcp)
+	if err != nil {
+		return nil, err
+	}
+
+	return md, nil
+}
+
+// Decodes a DriverCheckpointJson received from the Open txn request.
+func unmarshalDriverCheckpointJSON(raw json.RawMessage) (flowCheckpoint []byte, nextSeqNumList []int, err error) {
+	flowCheckpoint, nextSeqNumList, err = nil, nil, nil
+
+	if raw == nil {
+		return
+	}
+
+	var parsed driverCheckpoint
+	if err = pf.UnmarshalStrict(raw, &parsed); err != nil {
+		return
+	}
+
+	nextSeqNumList = parsed.NextSeqNumList
+	if flowCheckpoint, err = base64.StdEncoding.DecodeString(parsed.B64EncodedFlowCheckpoint); err != nil {
+		return
+	} else if flowCheckpoint != nil && len(flowCheckpoint) == 0 {
+		flowCheckpoint = nil
+	}
+
+	return
+}
+
+// driver implements the DriverServer interface.
+type driver struct{}
+
+func (driver) Spec(ctx context.Context, req *pm.SpecRequest) (*pm.SpecResponse, error) {
+	endpointSchema, err := jsonschema.Reflect(new(config)).MarshalJSON()
```

Comment:
Yes, in general the [post](https://stackoverflow.com/questions/30086302/what-is-the-difference-between-newt-and-t#:~:text=The%20two%20forms%20new(T,only%20do%20new(int)%20.) shows `new(T)` is equivalent to `&T{}`.
 
 While according to Uber's style guide
 ```
 Use &T{} instead of new(T) when initializing struct references so that it is consistent with the struct initialization.
```
 https://github.com/uber-go/guide/blob/master/style.md
 
 So I have made the change according to uber.